### PR TITLE
ZKGroups contract independent from protocols

### LINKED ISSUE #84: On-chain proof verification
**Description**:

The ZK-Groups contract should be independent from the protocol used to verify proofs. For each protocol a new specific contract can be added.

**Tasks**:

- [x] Make `ZKGroups.sol` contract independent
- [x] Create specific contract to verify proofs with Semaphore
- [x] Update `contract-utils` lib

### PR DESCRIPTION
**Pull request by cedoor**
## Description

This PR makes the `ZKGroups.sol` independent from the protocols. It can only be used to update the group fingerprints, that can then be used by specific protocols' contracts to verify zero-knowledge proofs on-chain.

## Related Issue

Closes #84 

## Does this introduce a breaking change?

-   [x] Yes
-   [ ] No

## Other information

Although the contracts of this PR are now Semaphore agnostic, the API services are still dependent on it. In the future, it will also be necessary to make those services (in particular `GroupsService`) independent of the protocols and data structures used for the groups, perhaps integrating a system to also allow ZKGroups to be extended with external plugins.

### CODE REVIEW

File: apps/contracts/contracts/IZKGroups.sol [hunk 1]
```diff
@@ -3,74 +3,22 @@ pragma solidity ^0.8.4;
...
-    error ZKGroups__YouAreUsingTheSameNullifierTwice();
-    error ZKGroups__MerkleTreeDepthIsNotSupported();
-    error ZKGroups__ParametersMustHaveTheSameLength();
-
-    struct GroupData {
-        uint256 merkleTreeRoot;
-        uint256 merkleTreeDepth;
+    struct Group {
+        uint256 id;
+        uint256 fingerprint;
```
Comment (cedoor) [inline apps/contracts/contracts/IZKGroups.sol:8]: @saleel Can `fingerprint` be the right name here? How does it sound?

File: apps/contracts/project.json [hunk 2]
```diff
@@ -8,7 +8,7 @@
             "options": {
                 "commands": [
                     "hardhat node",
-                    "hardhat compile && hardhat deploy --network localhost"
+                    "hardhat node hardhat compile && hardhat deploy-zkgroups --network localhost"
```
Comment (cedoor) [inline apps/contracts/project.json:11]: @saleel I noticed that this doesn't work on my side. The second command starts before the first one.

Comment (saleel) [reply apps/contracts/project.json:11]: @cedoor In this case we can add a delay before the second command. I had this initially but was working without as well for me.
Try this
```sleep 3 && hardhat compile && hardhat deploy-zkgroups --network localhost```

I don't understand your current change though - adding additional `hardhat node` before the second command?

**Review by saleel — APPROVED** [review]

File: apps/contracts/project.json (hunk 2 above)
Comment (cedoor) [reply apps/contracts/project.json]: Ops, just fixed! Thanks

**Review by saleel — APPROVED** [review]

**Pull request merged**